src/Associativity.cpp

```diff
@@ -0,0 +1,658 @@
+        }
+    }
+
+    void visit(const Call *op) {
+        if (!is_solvable) {
+            return;
+        }
+        if (op->call_type != Call::Halide) {
+            is_solvable = false;
+            return;
+        }
+
+        // Mutate the args
+        for (size_t i = 0; i < op->args.size(); i++) {
+            Expr new_args = mutate(op->args[i]);
+            if (type != OP_Y) {
+                is_solvable = false;
+                return;
+            }
+        }
+        internal_assert(type == OP_Y);
+        current_y = Expr(op);
+        expr = Variable::make(op->type, op_y);
+    }
+
+public:
+    ExtractBinaryOp(const string &f, const vector<Expr> &args, const map<int, Expr> &x_subs,
+                      const string &y, const map<Expr, string, ExprCompare> &y_subs) :
+        func(f), args(args), self_ref_subs(x_subs), op_y(y), y_subs(y_subs), is_solvable(true) {}
+
+    bool is_solvable;
+    Expr current_y;
+};
+
+// Given a binary expression operator 'bin_op' in the form of op(x, y), find
+// the identity of the operator. If the identity is not in [0, 1, +inf, -inf],
+// this returns undefined Expr
+Expr find_identity(Expr bin_op, const string &op_x, const string &op_y, Type t) {
+    vector<Expr> possible_identities = {make_const(t, 0), make_const(t, 1), t.min(), t.max()};
+    // For unary op (the one where 'x' does not appear), any value would be fine
+    for (const Expr &val : possible_identities) {
+        Expr subs = substitute(op_y, val, bin_op);
+        subs = common_subexpression_elimination(subs);
+        Expr compare = simplify(subs == Variable::make(t, op_x));
+        if (is_one(compare)) {
+            return val;
+        }
+    }
+    debug(4) << "Failed to find identity of " << bin_op << "\n";
+    return Expr(); // Fail to find the identity
+}
+
+// Given a binary expression operator 'bin_op' in the form of op(x, y), prove that
+// 'bin_op' is associative, i.e. prove that (x op y) op z == x op (y op z).
+// TODO(psuriana): The current implementation can't prove associativity of select()
+// as in argmax/argmin
+bool is_bin_op_associative(Expr bin_op, const string &op_x, const string &op_y, Type t) {
+    Expr x = Variable::make(t, op_x);
+    Expr y = Variable::make(t, op_y);
+    string op_z = unique_name("_z");
+    Expr z = Variable::make(t, op_z);
+
+    Expr lhs = substitute(op_y, z, bin_op);
+    lhs = substitute(op_x, bin_op, lhs);
+
+    Expr rhs = substitute({{op_x, y}, {op_y, z}}, bin_op);
+    rhs = substitute(op_y, rhs, bin_op);
+
+    // Canonicalize the lhs and rhs before comparing them so that we get
+    // a better chance of simplifying the equality.
+    vector<string> vars = {op_x, op_y, op_z};
+    for (const string &v : vars) {
+        lhs = solve_expression(lhs, v);
+        rhs = solve_expression(rhs, v);
+    }
+
+    lhs = common_subexpression_elimination(lhs);
+    rhs = common_subexpression_elimination(rhs);
+    Expr compare = simplify(lhs == rhs);
```

Comment:
Might be better to do common_subexpression_elimination on the equality test, rather than doing it individually on each side